Targeted fix for installing roles with symlinks containing '..'

s-hertel opened the pull request:
##### SUMMARY

Fixes https://github.com/ansible/ansible/issues/81965

I'd like to backport this as far as acceptable since the fix for the symlink CVE was backported to 2.13, so I separated this from the other fixes in https://github.com/ansible/ansible/pull/82052 related to installing broken roles.

##### ISSUE TYPE

<!--- Pick one below and delete the rest -->

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->

```paste below

```

s-hertel committed:
Targeted fix for installing roles with symlinks that contain '..'

s-hertel committed:
changelog

Akasurde commented:
/azp run

bcoca commented on lib/ansible/galaxy/role.py:
```diff
@@ -45,6 +45,20 @@
+
+        return not (path.startswith(archive_dir) and (not path_must_exist or os.path.exists(path)))
+    except OSError:
+        return True
```
lets not rewrite things, you can use `from lib/ansible/utils/path.py import is_subpath`

s-hertel replied:
didn't know it existed, thanks for pointing it out

s-hertel committed:
Refactor to use ansible.utils.path functions

Set the tarfile attribute to a normalized value from unfrackpath instead
of validating path parts and omiting potentially invald parts

Allow tarfile paths/links containing '..', '$', '~' as long as the
normalized realpath is in the tarfile's role directory

s-hertel commented:
I thought I might be able to simplify this quite a bit by using `tarfile.extractfile` to verify the member is in the tarfile before extracting:
```python
try:
    # throw KeyError if the member missing or external
    file_data = role_tar_file.extractfile(member)
except KeyError as e:
    raise AnsibleError(f"Invalid tarfile member is not a subpath of the role: {e}")

if file_data is None:
    # happens when not (member.isreg() or member.issym()) - we don't extract these
    continue

# then extract as usual
```
But `tarfile.extractfile` does not normalize the path before using it to determine if it exists in the tarfile, so this won't help.

s-hertel commented:
/azp run

bcoca commented on lib/ansible/galaxy/role.py:
```diff
@@ -392,43 +393,41 @@ def install(self):
+
+                                if attr_value.startswith(os.sep) and not is_subpath(attr_value, archive_parent_dir):
+                                    err = f"Invalid {attr} for tarfile member: path {attr_value} is not a subpath of the role {archive_parent_dir}"
+                                    raise AnsibleError(err)
```
should we add toggle to make this 'warn and skip' vs fatal error? 
do we even trust author at this point?

s-hertel replied:
I would say the tarfile is malicious at this point (unless my tarfile-fu is weak and I've missed an acceptable use-case that won't work with the fix). I'd be open to adding a toggle to allow skipping dangerous content because it's better than installing incomplete content without a warning, but I'm not sure when it would be useful.

s-hertel commented:
@bcoca thanks for the reviews
